### ISSUE TITLE
TestEntitlement CI Problem

### ISSUE THREAD

**Issue by xsalefter** [killbill/killbill#1694]
Sometimes in CI, `TestEntitlement#testCreateSubscriptionBillingInThePast()` has assertion error (that will described below). [This is the first PR](https://github.com/killbill/killbill/pull/1685#issuecomment-1120094414) that encounter this problem.

But when run this test in local using IDE, `TestEntitlement#testCreateSubscriptionBillingInThePast()` result is Ok and consistent without any problem. Running `TestEntitlement` in terminal using `mvn test -Dorg.killbill.billing.dbi.test.mysql=true -Dtest="TestEntitlement"` command is Ok without problem.

On the other side (but not sure if this is related to the main problem), running `TestEntitlement` test suite using IDE locally will lead to failed test in `TestEntitlement#testSubscriptionEventsWithCatalogChange()` (Log in comment)

Local test done both in linux (ubuntu) and macOS, using OpenJDK 11.

`TestEntitlement#testCreateSubscriptionBillingInThePast()` log:

```log
....
2022-05-17T07:57:25.4286915Z [INFO] 
2022-05-17T07:57:25.4287389Z [INFO] Results:
2022-05-17T07:57:25.4287914Z [INFO] 
2022-05-17T07:57:25.4290962Z [ERROR] Failures: 
2022-05-17T07:57:25.4291893Z [ERROR]   TestEntitlement>GuicyKillbillTestSuite.run:208->testCreateSubscriptionBillingInThePast:871 
expected 
[class Subscription {
     org.killbill.billing.client.model.gen.Subscription@dd8820f
     accountId: 63ed1e1d-b583-43e9-9123-fed802812c35
     bundleId: 9eb3e6b5-c34a-4322-95ea-8d16e556fb25
     bundleExternalKey: 9eb3e6b5-c34a-4322-95ea-8d16e556fb25
     subscriptionId: 75f4a971-67df-48a3-977a-a363b23e21e5
     externalKey: 75f4a971-67df-48a3-977a-a363b23e21e5
     startDate: 2012-04-25T00:03:42.000Z
     productName: Shotgun
     productCategory: BASE
     billingPeriod: MONTHLY
     phaseType: EVERGREEN
     priceList: DEFAULT
     planName: shotgun-monthly
     state: ACTIVE
     sourceType: NATIVE
     cancelledDate: null
     chargedThroughDate: 2012-03-25
     billingStartDate: 2012-03-25T00:03:42.000Z
     billingEndDate: null
     billCycleDayLocal: 24
     events: [class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@fa195e2d
         eventId: 4799b6a1-6ca0-40f1-ac84-1a21d635e8bf
         billingPeriod: MONTHLY
         effectiveDate: 2012-03-25T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: START_BILLING
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: billing-service
         serviceStateName: START_BILLING
         phase: shotgun-monthly-trial
         auditLogs: []
     }, class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@60d72475
         eventId: 5991af2f-9e4f-4a2d-9a78-607217e54d2d
         billingPeriod: MONTHLY
         effectiveDate: 2012-04-24T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: PHASE
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: entitlement+billing-service
         serviceStateName: PHASE
         phase: shotgun-monthly-evergreen
         auditLogs: []
     }, class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@2ee2b863
         eventId: d41683dd-86ae-4478-87d2-c598e082c0b6
         billingPeriod: MONTHLY
         effectiveDate: 2012-04-25T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: START_ENTITLEMENT
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: entitlement-service
         serviceStateName: ENT_STARTED
         phase: shotgun-monthly-evergreen
         auditLogs: []
     }]
     priceOverrides: null
     prices: [class PhasePrice {
         planName: shotgun-monthly
         phaseName: shotgun-monthly-trial
         phaseType: TRIAL
         fixedPrice: 0
         recurringPrice: null
         usagePrices: []
     }, class PhasePrice {
         planName: shotgun-monthly
         phaseName: shotgun-monthly-evergreen
         phaseType: EVERGREEN
         fixedPrice: null
         recurringPrice: 249.95
         usagePrices: []
     }]
     auditLogs: []
 }] but found 
 [class Subscription {
     org.killbill.billing.client.model.gen.Subscription@f694af38
     accountId: 63ed1e1d-b583-43e9-9123-fed802812c35
     bundleId: 9eb3e6b5-c34a-4322-95ea-8d16e556fb25
     bundleExternalKey: 9eb3e6b5-c34a-4322-95ea-8d16e556fb25
     subscriptionId: 75f4a971-67df-48a3-977a-a363b23e21e5
     externalKey: 75f4a971-67df-48a3-977a-a363b23e21e5
     startDate: 2012-04-25T00:03:42.000Z
     productName: Shotgun
     productCategory: BASE
     billingPeriod: MONTHLY
     phaseType: EVERGREEN
     priceList: DEFAULT
     planName: shotgun-monthly
     state: ACTIVE
     sourceType: NATIVE
     cancelledDate: null
     chargedThroughDate: 2012-05-24
     billingStartDate: 2012-03-25T00:03:42.000Z
     billingEndDate: null
     billCycleDayLocal: 24
     events: [class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@fa195e2d
         eventId: 4799b6a1-6ca0-40f1-ac84-1a21d635e8bf
         billingPeriod: MONTHLY
         effectiveDate: 2012-03-25T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: START_BILLING
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: billing-service
         serviceStateName: START_BILLING
         phase: shotgun-monthly-trial
         auditLogs: []
     }, class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@60d72475
         eventId: 5991af2f-9e4f-4a2d-9a78-607217e54d2d
         billingPeriod: MONTHLY
         effectiveDate: 2012-04-24T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: PHASE
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: entitlement+billing-service
         serviceStateName: PHASE
         phase: shotgun-monthly-evergreen
         auditLogs: []
     }, class EventSubscription {
         org.killbill.billing.client.model.gen.EventSubscription@2ee2b863
         eventId: d41683dd-86ae-4478-87d2-c598e082c0b6
         billingPeriod: MONTHLY
         effectiveDate: 2012-04-25T00:03:42.000Z
         plan: shotgun-monthly
         product: Shotgun
         priceList: DEFAULT
         eventType: START_ENTITLEMENT
         isBlockedBilling: false
         isBlockedEntitlement: false
         serviceName: entitlement-service
         serviceStateName: ENT_STARTED
         phase: shotgun-monthly-evergreen
         auditLogs: []
     }]
     priceOverrides: null
     prices: [class PhasePrice {
         planName: shotgun-monthly
         phaseName: shotgun-monthly-trial
         phaseType: TRIAL
         fixedPrice: 0
         recurringPrice: null
         usagePrices: []
     }, class PhasePrice {
         planName: shotgun-monthly
         phaseName: shotgun-monthly-evergreen
         phaseType: EVERGREEN
         fixedPrice: null
         recurringPrice: 249.95
         usagePrices: []
     }]
     auditLogs: []
 }]
....
```

**Comment by xsalefter** [comment]
IMO the problem is lied in line ~872 : `Assert.assertEquals(subscription, entitlementJson);`. 

Lets take a look at other assertion in `TestEntitlement`. If I understand `TestEntitlement#verifyChargedThroughDate()` correctly, it used to get up-to-date Subscription's `chargedThroughDate` value. In test case, `entitlementJson.chargedThroughDate()` value still `2012-03-24`, where the "correct" value is `2012-05-24`.

In short, in snippet below, there's no guarantee that,
```java 
final Subscription entitlementJson = subscriptionApi.createSubscription( ... ) // line ~849
```

The `entitlementJson` will return up-to-date or expected `chargedThroughDate` value.

With this in mind, back to first sentence of this comment, assertion in this line:
```java
Assert.assertEquals(subscription, entitlementJson);
```
will not always equal, because `equals()` method in `Subscription.java` (`killbill-client-java`) add `chargedThroughDate` to comparison (`Subscription.java` line 303). 

This is the diff of those 2 objects: https://www.diffchecker.com/uHlaknex

**Comment by pierre** [comment]
Yup, you're right! This comment explains it actually: https://github.com/killbill/killbill/blob/ea72fabd8389018bef3236ec6dd334f789f3be73/profiles/killbill/src/test/java/org/killbill/billing/jaxrs/TestEntitlement.java#L1419

I would remove `Assert.assertEquals(subscription, entitlementJson)` and replace it with a series of `assertEquals` statements on individual fields (all but the CTD).

**Comment by xsalefter** [comment]
Fixed by this commit: 081c08ca4994efcb56e0938ebc55f588d4281fcc
